### PR TITLE
feat: Support tracePropagationTargets

### LINKED ISSUE #2148: Support tracePropagationTargets
### Description

According to https://github.com/getsentry/team-mobile/issues/49, we need to add an option to users provide tracePropagationTargets in order to send trace_id header only to this targets.

See https://develop.sentry.dev/sdk/performance/#tracepropagationtargets for more information.

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

We now have a new SentryOption `tracePropagationTargets` which determines if our two headers are added to outgoing requests.

## :bulb: Motivation and Context

Closes #2148.

## :green_heart: How did you test it?

Unit test & running the app with a debug proxy to view actual requests and their headers.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by kevinrenskers** [comment]
Right now `SentrySDK.options.tracePropagationTargets` can contain either instances of `NSString`, or instances of `NSRegularExpression`.

In the case of `NSString`, the request's `URL.host` must match the value precisely. 
In the case of `NSRegularExpression`, the entire `URL` will be used for matching.

This matches the docs and example on https://develop.sentry.dev/sdk/performance/#tracepropagationtargets, and makes the most sense to me, and matches what I've seen in other systems (like CORS whitelists in Django). 

@brustolin on the other hand thinks that the entire url should always be used, and we should use `contains` rather than `isEqualToString`. I wanted to get early feedback on this. To me it doesn't really make sense, because if you add `sentry.io` to `SentrySDK.options.tracePropagationTargets`, what you mean is that all request from that domain are allowed. But `sentry.io.evil-hacker.com` would also match by using `contains`! If you really want to have that kind of freedom you can always use `NSRegularExpression`.

@philipphofmann @bruno-garcia what do you think?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (kevinrenskers) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:513]: I wonder if we still need the test-server, and those 2 tests that do an actual request to the test server. All we're really testing is if it headers are added, which we can test way easier like I added here.

File: test-server/Sources/App/routes.swift
Comment (kevinrenskers) [inline test-server/Sources/App/routes.swift:11]: `Hello, world! Trace header added.` seems way less useful than echoing the actual header, so we can test against that. I've modified the existing test that hits the test server, but I think we don't need the test server at all any more, see comment above.

**Comment by kevinrenskers** [comment]
I'm getting test failures in `testGetRequest_SpanCreatedAndBaggageHeaderAdded`, it seems that it uses a cached version of the server maybe? It all succeeds locally when I run the server with my changes. But if we're going to remove those tests that hit the server, I don't really have to look into any further. Let me know!

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:513]: Because the way we swizzle  NSSessionTask, we can't guarantee that the header injection is working on every iOS release, we still need to do a integration test.

Comment (kevinrenskers) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:513]: For which iOS versions is this needed? Is this something we can clean up with 8.0.0?

File: Sources/Sentry/SentryNetworkTracker.m
Comment (brustolin) [inline Sources/Sentry/SentryNetworkTracker.m:76]: h: Every other SDK uses contains and compare with the url not only the host.
We should align this.
```suggestion
            if ([URL.absoluteString containsString:targetCheck]) {
```

Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTracker.m:76]: That really doesn't make sense, and if other SDKs use this, they should fix it. See my comment in the PR about this enabling very weird things like sentry.io.evil.com getting headers when you check for sentry.io.

Not that these headers are very useful for evil hackers but I mean that's what this feature is for. Checking hostnames using contains is just fundamentally flawed. I'm waiting for input from Philip and Bruno, will bring this up in the meeting tonight.

**Comment by brustolin** [comment]
We are missing 'tracePropagationTargets' initialisation during `validateOptions:didFailWithError:` function in SentryOptions

**Comment by kevinrenskers** [comment]
> We are missing 'tracePropagationTargets' initialisation during `validateOptions:didFailWithError:` function in SentryOptions

Yeah I actually wanted to ask about that - what is the use of that `initWithDict` method anyways? Who's using this, and why?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (kevinrenskers) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:513]: Ah no because now we manually call `urlSessionTaskResume` and thus kicking off the test, but that skips the swizzling. Still feels a bit heavy handed, this test server just to test sending 2 headers that are not needed for Sentry itself to work. Oh well.

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift
Comment (brustolin) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift:161]: We should not use this anymore.

File: test-server/Sources/App/routes.swift
Comment (brustolin) [reply test-server/Sources/App/routes.swift:11]: I agree that echoing information is way more useful.

**Comment by philipphofmann** [comment]
> > Yeah I actually wanted to ask about that - what is the use of that initWithDict method anyways? Who's using this, and why?
> 
> I believe this is legacy stuff. @philipphofmann do you have a better answer?

Legacy and hybrid SDKs.

**Comment by philipphofmann** [comment]
@kevinrenskers, it's supposed to be as pointed out in the develop docs. We have an internal [DACI](https://www.notion.so/sentry/DACI-Control-Sending-Trace-Data-63fd04ea885443ee99514c15a1e91499), which I'm going to send you somewhere in private.

File: Sources/Sentry/SentryNetworkTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNetworkTracker.m:76]: We agreed to keep the behaviour used in the other SDKs and use `containsString` for the entire URL.

File: Sources/Sentry/SentryNetworkTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTracker.m:63]: `l`: We only need `URL.absoluteString` so we could only pass the `absoluteString`.

File: Sources/Sentry/SentryTraceContext.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTraceContext.m:115]: `l`: It would be nice to have such changes in a separate PR. It makes reviewing this easier.

File: Sources/Sentry/SentryNetworkTrackingIntegration.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTrackingIntegration.m:77]: `h`: Why don't we need this anymore? Fantastic, if we can get rid of some swizzling.

File: Sources/Sentry/SentryNetworkTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTracker.m:167]: `m`: I think we don't need to add the header if it's empty, which could happen if the `tracer == nil` above.

File: Sources/Sentry/SentryNetworkTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTracker.m:321]: `m`: Why can we ditch that code?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (philipphofmann) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:513]: Using swizzling has a risk of breaking actual HTTP requests. We still use swizzling for HTTP requests. I would rather keep one integration test in combination with the test server. We don't have to test all the functionality with the test server.

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:544]: `m`: Shouldn't they be `nil`?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:496]: `m`: I think those two test cases are great. Why not keep them with the new code?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:547]: `l`: Excellent test case, but I think we should split it up into several ones.

**Review by philipphofmann — COMMENTED** [review]
Thanks for taking care of this, @kevinrenskers.

File: Sources/Sentry/Public/SentryOptions.h
Comment (kevinrenskers) [reply Sources/Sentry/Public/SentryOptions.h:409]: https://github.com/getsentry/sentry-docs/pull/5574

File: Sources/Sentry/SentryNetworkTracker.m
Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTracker.m:63]: Passing in an actual `NSURL` rather than an `NSString` seems more clear and obvious to me, just by looking at the types you see what to expect.

File: Sources/Sentry/SentryNetworkTracker.m
Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTracker.m:321]: I thought that we wouldn't run into the same problem with the wrong headers anymore but looking at it again, we can still run into the problem where NSURLSession itself has cached the old headers. Restoring this functionality, will add tests after lunch (I need to run an errand now).

Comment (brustolin) [reply Sources/Sentry/SentryNetworkTracker.m:321]: Now we are injecting the header directly into NSURLRequest, the session will not cache this.

Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTracker.m:321]: Right, that was also my first instinct but Philip's questions made me question myself 😅

I've removed it yet again and tested it, and indeed we don't need this workaround anymore.

File: Sources/Sentry/SentryOptions.m
Comment (kevinrenskers) [reply Sources/Sentry/SentryOptions.m:111]: This is tested using `testAddHeadersForRequestWithURL`. What are you missing?

File: Sources/Sentry/SentryNetworkTracker+Private.h
Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTracker+Private.h]: Good point, I must have thought it was public.

File: Sources/Sentry/SentryNetworkTrackingIntegration.m
Comment (kevinrenskers) [reply Sources/Sentry/SentryNetworkTrackingIntegration.m:77]: Because the only thing that we did in there (adding the baggage header) we no longer do :)

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (kevinrenskers) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:547]: Just wondering why you think so? Splitting it up doesn't really do anything, and I like how all these related things are now in one test function.

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift
Comment (kevinrenskers) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:496]: Done

**Review by brustolin — APPROVED** [review]
LGTM

Just wait for @philipphofmann replies in his comments before merge this.

File: Sources/Sentry/Public/SentryOptions.h
Comment (philipphofmann) [inline Sources/Sentry/Public/SentryOptions.h:409]: `m`: I think we should not point to the node docs, we should point to the apple docs.

**Review by philipphofmann — COMMENTED** [review]
Only missing tests and the wrong link to the docs.

File: Sources/Sentry/SentryOptions.m
Comment (kevinrenskers) [reply Sources/Sentry/SentryOptions.m:111]: > That the default is an array with a regex .*.

That is basically what is tested using `testAddHeadersForRequestWithURL`. Not that it literally is a regex `.*`, but that it has the correct behavior, which seems more useful to me than testing if it contains this specific regex?

I'll look into the other points.

File: Tests/SentryTests/SentryOptionsTest.m
Comment (philipphofmann) [inline Tests/SentryTests/SentryOptionsTest.m:278]: `m`: Let's please also validate the default value. We only need to add a line somewhere here https://github.com/getsentry/sentry-cocoa/blob/5ecf9f6f92ee2598fd21b95d1a3af356d9969c50/Tests/SentryTests/SentryOptionsTest.m#L519

**Review by philipphofmann — APPROVED** [review]
Thanks, 🙏

**Pull request merged**
